launch/launch/launch_service.py

```diff
@@ -338,28 +339,39 @@ def _on_exception(loop, context):
                         assert ret is None, ret
                         continue
-                    process_one_event_task = this_loop.create_task(self._process_one_event())
-                    if self.__shutting_down:
-                        # If shutting down and idle then we're done.
-                        if is_idle:
+
+                    # Stop running if we're shutting down and there's no more work
+                    if self.__shutting_down and is_idle:
+                        if (
+                            process_one_event_task is not None and
+                            not process_one_event_task.done()
+                        ):
                             process_one_event_task.cancel()
+                        break
+
+                    # Collect futures to wait on
+                    entity_futures = [pair[1] for pair in self._entity_future_pairs]
+                    entity_futures.extend(self.__context._completion_futures)
+
+                    # If the current task is done, create a new task to process any events
+                    # in the queue
+                    if process_one_event_task is None or process_one_event_task.done():
+                        process_one_event_task = this_loop.create_task(self._process_one_event())
+
+                    # Add the process event task to the list of awaitables
+                    entity_futures.append(process_one_event_task)
+
+                    # Wait on events and futures until an event is processed or a timeout occurs
+                    while True:
+                        done, pending = await asyncio.wait(
+                            entity_futures,
+                            timeout=1.0,
+                            return_when=asyncio.FIRST_COMPLETED
+                        )
+                        if (process_one_event_task in done) or len(done) == 0:
                             break
-                        entity_futures = [pair[1] for pair in self._entity_future_pairs]
-                        entity_futures.append(process_one_event_task)
-                        entity_futures.extend(self.__context._completion_futures)
-                        done = set()  # type: Set[asyncio.Future]
-                        while not done:
-                            done, pending = await asyncio.wait(
-                                entity_futures,
-                                timeout=1.0,
-                                return_when=asyncio.FIRST_COMPLETED
-                            )
-                            if not done:
-                                self.__logger.debug(
-                                    'still waiting on futures: {}'.format(entity_futures)
-                                )
-                    else:
-                        await process_one_event_task
+                        entity_futures = set(entity_futures).difference(done)
```

Comment:
can you use pending instead of doing this?
```suggestion
                        entity_futures = pending
```

Reply:
Removed the inner loop (b002b92), so this does not apply anymore.